pkg/asset/kubeconfig/kubeconfig_test.go

```diff
@@ -0,0 +1,229 @@
+package kubeconfig
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/stretchr/testify/assert"
+)
+
+type fakeAsset int
+
+var _ asset.Asset = fakeAsset(0)
+
+func (f fakeAsset) Dependencies() []asset.Asset {
+	return nil
+}
+
+func (f fakeAsset) Generate(map[asset.Asset]*asset.State) (*asset.State, error) {
+	return nil, nil
+}
+
+func TestKubeconfigGenerate(t *testing.T) {
+	testDir, err := ioutil.TempDir(os.TempDir(), "kubeconfig_test")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(testDir)
+
+	rootCA := fakeAsset(0)
+	adminCertKey := fakeAsset(1)
+	kubeletCertKey := fakeAsset(2)
+	installConfig := fakeAsset(3)
+
+	rootCAState := &asset.State{
+		Contents: []asset.Content{
+			{
+				Name: "root-ca.key",
+				Data: []byte("THIS IS ROOT CA KEY DATA"),
+			},
+			{
+				Name: "root-ca.crt",
+				Data: []byte("THIS IS ROOT CA CERT DATA"),
+			},
+		},
+	}
+
+	adminCertState := &asset.State{
+		Contents: []asset.Content{
+			{
+				Name: "admin.key",
+				Data: []byte("THIS IS ADMIN KEY DATA"),
+			},
+			{
+				Name: "admin.crt",
+				Data: []byte("THIS IS ADMIN CERT DATA"),
+			},
+		},
+	}
+
+	kubeletCertState := &asset.State{
+		Contents: []asset.Content{
+			{
+				Name: "kubelet.key",
+				Data: []byte("THIS IS KUBELET KEY DATA"),
+			},
+			{
+				Name: "kubelet.crt",
+				Data: []byte("THIS IS KUBELET CERT DATA"),
+			},
+		},
+	}
+
+	installConfigState := &asset.State{
+		Contents: []asset.Content{
+			{
+				Name: "installconfig.yaml",
+				Data: []byte(`baseDomain: test.example.com
+metadata:
+  name: test-cluster-name
+`),
+			},
+		},
+	}
+
+	tests := []struct {
+		userName     string
+		certKey      asset.Asset
+		parents      map[asset.Asset]*asset.State
+		errString    string
+		expectedData []byte
+	}{
+		{
+			userName: "admin",
+			certKey:  adminCertKey,
+			parents: map[asset.Asset]*asset.State{
+				rootCA:        rootCAState,
+				adminCertKey:  adminCertState,
+				installConfig: installConfigState,
+			},
+			errString: "",
+			expectedData: []byte(`clusters:
+- cluster:
+    certificate-authority-data: VEhJUyBJUyBST09UIENBIENFUlQgREFUQQ==
+    server: https://test-cluster-name-api.test.example.com:6443
+  name: test-cluster-name
+contexts:
+- context:
+    cluster: test-cluster-name
+    user: admin
+  name: admin
+current-context: admin
+preferences: {}
```

Comment:
nit: Maybe `preferences` should be `omitempty` upstream.

Reply:
:+1: do you want to open up a PR for that?

Reply:
> do you want to open up a PR for that?

Ah, `Config.Preferences` [is a struct upstream][1], so `omitempty` would currently be ignored (golang/go#11939).  To drop this useless `{}` we'd need to either fix the Go bug or convert `Config.Preferences` to a pointer (`*Preferences`).  Both of those approaches seem like more work than this warrants, so I'll just live with the `{}` ;).

[1]: https://github.com/kubernetes/client-go/blob/87935b98dd4abdf6477f29baf934d6dd0aba82d4/tools/clientcmd/api/v1/types.go#L37-L38